### PR TITLE
feat: support for all solver backends for cryptol server/client

### PR DESCRIPTION
**Pull request by pnwamk**
Makes available both SBV and What4 based solver backends in the RPC server and Python client.

Addresses https://github.com/GaloisInc/argo/issues/173

### CODE REVIEW

File: cryptol-remote-api/src/CryptolServer/Sat.hs [hunk 1]
```diff
@@ -124,25 +215,25 @@ instance ToJSON ProveSatResult where
                   | res <- xs
                   ]
                 ]
+  toJSON (OfflineSMTQuery smtlib) =
+    JSON.object [ "result" .= ("offline" :: Text)
```
Comment (RyanGlScott) [inline cryptol-remote-api/src/CryptolServer/Sat.hs:219]: Don't forget to update the `resultFieldDescription` with this new information.

File: cryptol-remote-api/python/cryptol/solver.py [hunk 2]
```diff
@@ -25,5 +53,8 @@
 W4_YICES: Solver     = Solver("w4-yices")
 W4_Z3: Solver        = Solver("w4-z3")
 W4_BOOLECTOR: Solver = Solver("w4-boolector")
-W4_MATHSAT: Solver   = Solver("w4-offline")
+W4_OFFLINE: Solver   = Solver("w4-offline")
 W4_ABC: Solver       = Solver("w4-any")
+
```
Comment (RyanGlScott) [inline cryptol-remote-api/python/cryptol/solver.py:58]: Trailing whitespace

File: cryptol-remote-api/src/CryptolServer/Sat.hs (hunk 1 above)
Comment (pnwamk) [reply cryptol-remote-api/src/CryptolServer/Sat.hs:219]: Thanks!

File: cryptol-remote-api/src/CryptolServer/Sat.hs [hunk 3]
```diff
@@ -27,76 +32,162 @@ import Cryptol.ModuleSystem (checkExpr)
...
+        Right smtlib -> pure $ OfflineSMTQuery smtlib
+    Right w4Cfg -> do
+      smtlibRef <- liftIO $ newIORef ""
+      result <- liftModuleCmd $
+        W4.satProveOffline w4Cfg hConsing False cmd $ \f -> do
+          withRWTempFile "smtOutput.tmp" $ \h -> do
+            f h
+            hSeek h AbsoluteSeek 0
+            contents <- TIO.hGetContents h
+            writeIORef smtlibRef $ T.unpack contents
```
Comment (atomb) [inline cryptol-remote-api/src/CryptolServer/Sat.hs:133]: Maybe don't unpack and make the field of `OfflineSMTQuery` a `Text` value instead? The contents of these files can get large sometimes.

File: cryptol-remote-api/src/CryptolServer/Sat.hs [hunk 4]
```diff
@@ -27,76 +32,162 @@ import Cryptol.ModuleSystem (checkExpr)
...
+        Nothing -> do
+          smtlib <- liftIO $ readIORef smtlibRef
+          pure $ OfflineSMTQuery smtlib
   where
-    satResult :: [(TValue, Expr, Value)] -> CryptolCommand [(JSONType, Expression)]
-    satResult es = traverse result es
+    withRWTempFile name k =
+      X.bracket
+        (do tmp <- getTemporaryDirectory
+            let esc c = if isAscii c && isAlphaNum c then c else '_'
```
Comment (atomb) [inline cryptol-remote-api/src/CryptolServer/Sat.hs:143]: Well done. We should have similar escaping mechanisms in other places.

File: cryptol-remote-api/src/CryptolServer/Sat.hs [hunk 5]
```diff
@@ -27,76 +32,162 @@ import Cryptol.ModuleSystem (checkExpr)
...
+          me <- readBack t v
+          case me of
+            Nothing -> error $ "type is not convertable: " ++ (show t)
+            Just e -> pure (JSONType mempty (tValTy t), e)
 
 data ProveSatResult
   = Unsatisfiable
   | Invalid CounterExampleType [(JSONType, Expression)]
   | Satisfied [[(JSONType, Expression)]]
+  | OfflineSMTQuery String
```
Comment (atomb) [inline cryptol-remote-api/src/CryptolServer/Sat.hs:190]: See comment about `T.unpack` above.

File: cryptol-remote-api/src/CryptolServer/Sat.hs [hunk 6]
```diff
@@ -124,25 +215,25 @@ instance ToJSON ProveSatResult where
                   | res <- xs
                   ]
                 ]
+  toJSON (OfflineSMTQuery smtlib) =
+    JSON.object [ "result" .= ("offline" :: Text)
+                , "query" .= (T.pack smtlib)
```
Comment (atomb) [inline cryptol-remote-api/src/CryptolServer/Sat.hs:220]: Won't need to do `T.pack` if the file contents are already `Text`.

**Review by atomb — CHANGES_REQUESTED** [review]
I think the offline SMT-Lib contents should probably be `Text` instead of `String`, but otherwise this is great. Thanks for updating the docs along with the code. :)

File: cryptol-remote-api/src/CryptolServer/Sat.hs (hunk 4 above)
Comment (pnwamk) [reply cryptol-remote-api/src/CryptolServer/Sat.hs]: I remembered I borrowed it from the REPL code -- we now import that to reuse it. We can import it wherever else we want.

**Comment by pnwamk** [comment]
Comments addressed and passing all tests (the two failures are unrelated). Planning on merging soon.

**Review by RyanGlScott — APPROVED** [review]

**Pull request merged**